head_test.go

```diff
@@ -161,3 +163,219 @@ func TestNonDuplicateNaNDatapointsCausesAmendError(t *testing.T) {
+			v := rand.Float64()
+			series = append(series, sample{ts, v})
+
+			_, err := app.Add(ls, ts, v)
+			require.NoError(t, err)
+
+			ts += rand.Int63n(timeInterval) + 1
+		}
+
+		seriesMap[labels.New(l...).String()] = series
+	}
+
+	require.NoError(t, app.Commit())
+
+	// Query each selector on 1000 random time-ranges.
+	queries := []struct {
+		ms []labels.Matcher
+	}{
+		{
+			ms: []labels.Matcher{labels.NewEqualMatcher("a", "b")},
+		},
+		{
+			ms: []labels.Matcher{
+				labels.NewEqualMatcher("a", "b"),
+				labels.NewEqualMatcher("job", "prom-k8s"),
+			},
+		},
+		{
+			ms: []labels.Matcher{
+				labels.NewEqualMatcher("a", "c"),
+				labels.NewEqualMatcher("instance", "localhost:9090"),
+				labels.NewEqualMatcher("job", "prometheus"),
+			},
+		},
+		// TODO: Add Regexp Matchers.
+	}
+
+	for _, qry := range queries {
+		matched := []labels.Labels{}
+	Outer:
+		for _, ls := range lbls {
+			for _, m := range qry.ms {
+				if !matchLSet(m, ls) {
+					continue Outer
+				}
+			}
+
+			matched = append(matched, ls)
+		}
+
+		sort.Slice(matched, func(i, j int) bool {
+			return labels.Compare(matched[i], matched[j]) < 0
+		})
+
+		for i := 0; i < numRanges; i++ {
+			mint := rand.Int63n(300)
+			maxt := mint + rand.Int63n(timeInterval*int64(numDatapoints))
+
+			q := hb.Querier(mint, maxt)
+			ss := q.Select(qry.ms...)
+
+			// Build the mockSeriesSet.
+			matchedSeries := make([]Series, 0, len(matched))
+			for _, m := range matched {
+				smpls := boundedSamples(seriesMap[m.String()], mint, maxt)
+
+				// Only append those series for which samples exist as mockSeriesSet
+				// doesn't skip series with no samples.
+				// TODO: But sometimes SeriesSet returns an empty SeriesIterator
+				if len(smpls) > 0 {
+					matchedSeries = append(matchedSeries, newSeries(
+						m.Map(),
+						smpls,
+					))
+				}
+			}
+			expSs := newListSeriesSet(matchedSeries)
+
+			// Compare both SeriesSets.
+			for {
+				eok, rok := expSs.Next(), ss.Next()
+
+				// HACK: Skip a series if iterator is empty.
+				if rok {
+					for !ss.At().Iterator().Next() {
+						rok = ss.Next()
+						if !rok {
+							break
+						}
+					}
+				}
+
+				require.Equal(t, eok, rok, "next")
+
+				if !eok {
+					break
+				}
+				sexp := expSs.At()
+				sres := ss.At()
+
+				require.Equal(t, sexp.Labels(), sres.Labels(), "labels")
+
+				smplExp, errExp := expandSeriesIterator(sexp.Iterator())
+				smplRes, errRes := expandSeriesIterator(sres.Iterator())
+
+				require.Equal(t, errExp, errRes, "samples error")
+				require.Equal(t, smplExp, smplRes, "samples")
+			}
+		}
+	}
+
+	return
+}
+
+func boundedSamples(full []sample, mint, maxt int64) []sample {
+	start, end := -1, -1
+	for i, s := range full {
+		if s.t >= mint {
+			start = i
+			break
+		}
+	}
+	if start == -1 {
+		start = len(full)
+	}
+
+	for i, s := range full[start:] {
+		if s.t > maxt {
+			end = start + i
+			break
+		}
+	}
+
+	if end == -1 {
+		end = len(full)
+	}
+
+	return full[start:end]
+}
```

Comment:
This is the culprit of randomized tests. This function extracts what the correct result data is but is so complex in itself, that it would need its own tests again. Many changes for off-by-one errors here and hard to verify by just looking at the code.

If we could make it simpler to the point where it is "obviously correct", that would be good.
Does this roughly implement the same behavior?

```go
// Truncate samples in the front that are before mint.
for len(full) > 0 {
    if full[0] >= mint {
        break
    }
    full = full[1:]
}
for i, s := range full {
    // Terminate on the first sample larger than maxt.
    if s.t > maxt {
        return full[:i]
    }
}
// maxt is after highest sample.
return full
```